File: frontend/src/api/macro.ts
import axios from 'axios';
import type { DashboardData, KospiBottomData, NasdaqBottomData } from '../types/macro';

const API_BASE = import.meta.env.VITE_API_URL || '/api';

const api = axios.create({
  baseURL: API_BASE,
  timeout: 120000, // 콜드 캐시 첫 수집(외부 API 병렬 호출)이 느린 날 대비 여유

});

/**
 * KST 06:00 기준 일간 인메모리 캐시
 * 백엔드도 동일 기준으로 캐시하지만, 프론트에서도 캐시하여
 * 페이지 전환 시 불필요한 네트워크 요청을 제거
 */
interface CacheEntry<T> {
  data: T;
  cachedAt: number; // timestamp ms
}

const cache: {
  dashboard: CacheEntry<DashboardData> | null;
  categories: Record<string, CacheEntry<Record<string, unknown>>>;
  signalHistory: CacheEntry<unknown[]> | null;
  kospiBottom: CacheEntry<KospiBottomData> | null;
  nasdaqBottom: CacheEntry<NasdaqBottomData> | null;
} = {
  dashboard: null,
  categories: {},
  signalHistory: null,
  kospiBottom: null,
  nasdaqBottom: null,
};

function getResetBoundary(): number {
  const now = new Date();
  // KST = UTC+9
  const kstHour = (now.getUTCHours() + 9) % 24;
  const kstDate = new Date(now.getTime() + 9 * 60 * 60 * 1000);
  const reset = new Date(kstDate);
  reset.setUTCHours(6 - 9, 0, 0, 0); // 06:00 KST = 21:00 UTC (prev day)
  if (kstHour < 6) {
    reset.setUTCDate(reset.getUTCDate() - 1);
  }
  return reset.getTime();
}

function isCacheValid<T>(entry: CacheEntry<T> | null): entry is CacheEntry<T> {
  if (!entry) return false;
  return entry.cachedAt >= getResetBoundary();
}

export const fetchDashboard = async (): Promise<DashboardData> => {
  if (isCacheValid(cache.dashboard)) return cache.dashboard.data;

  const response = await api.get('/macro/dashboard');
  cache.dashboard = { data: response.data, cachedAt: Date.now() };
  return response.data;
};

export const fetchCategoryDetail = async (category: string): Promise<Record<string, unknown>> => {
  const entry = cache.categories[category];
  if (isCacheValid(entry)) return entry.data;

  const response = await api.get(`/macro/category/${category}`);
  cache.categories[category] = { data: response.data, cachedAt: Date.now() };
  return response.data;
};

export const fetchSignalHistory = async () => {
  if (isCacheValid(cache.signalHistory)) return cache.signalHistory.data;

  const response = await api.get('/macro/signals/history');
  cache.signalHistory = { data: response.data, cachedAt: Date.now() };
  return response.data;
};

export const setElliottCount = async (count: number) => {
  const response = await api.post('/macro/elliott', { count });
  return response.data;
};

export const setCapexCompany = async (
  company: string,
  status: 'up' | 'flat' | 'down',
) => {
  const response = await api.post('/macro/capex/company', { company, status });
  cache.dashboard = null; // 캐시 무효화 → 다음 fetch에서 갱신
  return response.data;
};

export const setDram = async (yoy: number, momentum: 'accel' | 'decel') => {
  const response = await api.post('/macro/capex/dram', { yoy, momentum });
  cache.dashboard = null;
  return response.data;
};

export const fetchKospiBottom = async (force = false): Promise<KospiBottomData> => {
  if (!force && isCacheValid(cache.kospiBottom)) return cache.kospiBottom.data;
  const response = await api.get('/macro/kospi-bottom', {
    params: force ? { refresh: true } : undefined,
  });
  cache.kospiBottom = { data: response.data, cachedAt: Date.now() };
  return response.data;
};

export const fetchNasdaqBottom = async (): Promise<NasdaqBottomData> => {
  if (isCacheValid(cache.nasdaqBottom)) return cache.nasdaqBottom.data;
  const response = await api.get('/macro/nasdaq-bottom');
  cache.nasdaqBottom = { data: response.data, cachedAt: Date.now() };
  return response.data;
};

export const setKospiManual = async (
  credit: 'rising' | 'falling' | 'stalling',
  forced: 'spike' | 'normal' | 'easing',
) => {
  const response = await api.post('/macro/kospi/manual', { credit, forced });
  cache.kospiBottom = null;
  return response.data;
};

File: backend/app/routers/macro.py
"""매크로 분석 API 라우터"""
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from ..services.macro_service import macro_service, VALID_CATEGORIES

router = APIRouter(prefix="/api/macro", tags=["Macro"])


class ElliottInput(BaseModel):
    """엘리엇 파동 수동 입력"""
    count: int = Field(..., ge=0, le=3, description="엘리엇 5파동 연속 출현 횟수 (0~3)")


class CapexCompanyInput(BaseModel):
    """빅테크 캐펙스 가이던스 수동 입력 (핵심 판별자)"""
    company: str = Field(..., description="alphabet | microsoft | meta | amazon | tsmc | broadcom")
    status: str = Field(..., pattern="^(up|flat|down)$", description="up(상향/유지) | flat(중립) | down(둔화/하향)")


class DramInput(BaseModel):
    """D램 가격 상승률 수동 입력 (엔진 기울기)"""
    yoy: float = Field(..., description="D램 가격 상승률 % (예: 15.0)")
    momentum: str = Field(..., pattern="^(accel|decel)$", description="accel(가속) | decel(감속)")


class KospiManualInput(BaseModel):
    """코스피 저점 수동 입력 (신용잔고 추이 + 반대매매)"""
    credit: str = Field(..., pattern="^(rising|falling|stalling)$", description="신용잔고: rising(증가) | falling(청산중) | stalling(멈춤)")
    forced: str = Field(..., pattern="^(spike|normal|easing)$", description="반대매매: spike(급증) | normal | easing(진정)")


@router.get("/dashboard")
async def get_dashboard():
    """매크로 대시보드 - 종합 판정 + 카테고리 요약 + 시그널"""
    try:
        return macro_service.get_dashboard()
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Dashboard failed: {str(e)}")


@router.get("/category/{name}")
async def get_category_detail(name: str):
    """카테고리별 상세 데이터"""
    if name not in VALID_CATEGORIES:
        raise HTTPException(status_code=400, detail=f"Invalid category: {name}. Valid: {VALID_CATEGORIES}")

    try:
        return macro_service.get_category_detail(name)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Category fetch failed: {str(e)}")


@router.get("/signals/history")
async def get_signal_history():
    """시그널 상태 변경 이력"""
    return macro_service.get_signal_history()


@router.post("/elliott")
async def set_elliott(input: ElliottInput):
    """엘리엇 파동 수동 입력 (0~3)"""
    return macro_service.set_elliott_count(input.count)


@router.post("/capex/company")
async def set_capex_company(input: CapexCompanyInput):
    """빅테크 캐펙스 가이던스 수동 입력"""
    try:
        return macro_service.set_capex_company(input.company, input.status)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/capex/dram")
async def set_dram(input: DramInput):
    """D램 가격 상승률 수동 입력"""
    return macro_service.set_dram(input.yoy, input.momentum)


@router.get("/kospi-bottom")
async def get_kospi_bottom(refresh: bool = False):
    """코스피 저점 판정 - 파라볼릭 되돌림 + 낙폭 밴드 + 신용/반대매매 (refresh=true 시 즉시 재취득)"""
    try:
        return macro_service.get_kospi_bottom(force=refresh)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"KOSPI bottom failed: {str(e)}")


@router.get("/nasdaq-bottom")
async def get_nasdaq_bottom(refresh: bool = False):
    """나스닥 저점 판정 - 파라볼릭 되돌림 + 낙폭 밴드 + 역대 약세장"""
    try:
        return macro_service.get_nasdaq_bottom(force=refresh)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"NASDAQ bottom failed: {str(e)}")


@router.post("/kospi/manual")
async def set_kospi_manual(input: KospiManualInput):
    """코스피 저점 수동 입력 (신용잔고 + 반대매매)"""
    try:
        return macro_service.set_kospi_manual(input.credit, input.forced)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/kofia/test")
async def kofia_test():
    """KOFIA 신용공여잔고 OpenAPI 연결 테스트 (키 발급 후 필드명 확인용)"""
    from ..services.kofia_fetcher import kofia_fetcher
    if not kofia_fetcher.enabled:
        return {"enabled": False, "hint": "KOFIA_API_KEY(.env) 미설정"}
    try:
        return {"enabled": True, "raw": kofia_fetcher.raw_sample(), "parsed": kofia_fetcher.get_credit_balance()}
    except Exception as e:
        raise HTTPException(status_code=502, detail=f"KOFIA call failed: {str(e)}")


@router.get("/customs/test")
async def customs_test():
    """관세청 반도체 수출 API 연결 테스트 (활용신청 반영 확인용)"""
    from ..services.customs_export_fetcher import customs_export_fetcher
    if not customs_export_fetcher.enabled:
        return {"enabled": False, "hint": "KOFIA_API_KEY(.env) 미설정"}
    return {"enabled": True, "raw": customs_export_fetcher.raw_sample(),
            "parsed": customs_export_fetcher.get_semiconductor_export()}


@router.get("/debug/kitchin")
async def debug_kitchin():
    """키친사이클 디버그 - 각 지표 트렌드 상세"""
    try:
        return macro_service.debug_kitchin_cycle()
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Debug failed: {str(e)}")
